From dd16c35412afae1cd7a3c036ddd7f7ec1b9aa40b Mon Sep 17 00:00:00 2001
From: MohammadHossein Qanbari <mohammad.qanbari@qt.io>
Date: Wed, 26 Mar 2025 13:51:11 +0100
Subject: Basic Style: Add support for dark mode color scheme

Previously, the basic style did not respect the system's dark mode
settings because the palette initialized by QQuickBasicTheme lacked
definitions for dark mode colors.

This commit updates the palette initialization to dynamically set colors
based on the current system color scheme. Additionally, the palette is
reinitialized whenever the system color scheme changes.

The controls gallery example has been updated to demonstrate the new
functionality.

[ChangeLog][Controls][Basic] Basic style supports dark mode now.

Fixes: QTBUG-135207
Change-Id: I2d6a74b407a7981905a9b86e97004cf0609a4bf0
Reviewed-by: Mitch Curtis <mitch.curtis@qt.io>
(cherry picked from commit 52141b34ecd16b33e9879fc1047c6f9c3dfde6da)
---
 examples/quickcontrols/gallery/gallery.qml    |  2 +-
 src/quickcontrols/basic/qquickbasictheme.cpp  | 84 +++++++++++++------
 src/quickcontrols/basic/qquickbasictheme_p.h  |  1 +
 .../qtquickcontrols2basicstyleplugin.cpp      |  6 ++
 4 files changed, 68 insertions(+), 25 deletions(-)

diff --git x/qtdeclarative/examples/quickcontrols/gallery/gallery.qml y/qtdeclarative/examples/quickcontrols/gallery/gallery.qml
index 15d49fa52e3898cdd607dc6e499beca37645ac25..3481fc9a151ef902df82984dd0f4ad3d144fe368 100644
--- x/qtdeclarative/examples/quickcontrols/gallery/gallery.qml
+++ y/qtdeclarative/examples/quickcontrols/gallery/gallery.qml
@@ -271,7 +271,7 @@ ApplicationWindow {
                 id: colorSchemes
                 // Some Qt Quick styles prioritize the respective design system guidelines
                 // over the system palette.
-                enabled: ["FluentWinUI3", "Fusion", "iOS"].includes(styleBox.currentText)
+                enabled: ["FluentWinUI3", "Fusion", "iOS", "Basic"].includes(styleBox.currentText)
                 CheckBox {
                     id: autoColorScheme
                     checked: true
diff --git x/qtdeclarative/src/quickcontrols/basic/qquickbasictheme.cpp y/qtdeclarative/src/quickcontrols/basic/qquickbasictheme.cpp
index 7884ad102e98072f49b8d7647e06b2248fdebe0f..735b4e95690f176e103e8d104013b225fae55f26 100644
--- x/qtdeclarative/src/quickcontrols/basic/qquickbasictheme.cpp
+++ y/qtdeclarative/src/quickcontrols/basic/qquickbasictheme.cpp
@@ -4,6 +4,7 @@
 #include "qquickbasictheme_p.h"
 
 #include <QtQuickTemplates2/private/qquicktheme_p.h>
+#include <QtQuickControls2/private/qquickstyle_p.h>
 
 QT_BEGIN_NAMESPACE
 
@@ -11,48 +12,83 @@ void QQuickBasicTheme::initialize(QQuickTheme *theme)
 {
     QPalette systemPalette;
 
-    systemPalette.setColor(QPalette::Base, QColor::fromRgba(0xFFFFFFFF));
-    systemPalette.setColor(QPalette::Disabled, QPalette::Base, QColor::fromRgba(0xFFD6D6D6));
+    const bool isDarkSystemTheme = QQuickStylePrivate::isDarkSystemTheme();
 
-    systemPalette.setColor(QPalette::Button, QColor::fromRgba(0xFFE0E0E0));
+    const QRgb base(isDarkSystemTheme ? 0xFF000000 : 0xFFFFFFFF);
+    const QRgb disabledBase(isDarkSystemTheme ? 0xFF292929 : 0xFFD6D6D6);
+    const QRgb button(isDarkSystemTheme ? 0xFF2F2F2F : 0xFFE0E0E0);
+    const QRgb buttonText(isDarkSystemTheme ? 0xFFD4D6D8 : 0xFF26282A);
+    const QRgb disabledButtonText(isDarkSystemTheme ? 0x4DD4D6D8 : 0x4D26282A);
+    const QRgb brightText(isDarkSystemTheme ? 0xFF000000 : 0xFFFFFFFF);
+    const QRgb disabledBrightText(isDarkSystemTheme ? 0x4D000000 : 0x4DFFFFFF);
+    const QRgb dark(isDarkSystemTheme ? 0xFFC8C9CB : 0xFF353637);
+    const QRgb highlight(isDarkSystemTheme ? 0xFF0D69F2 : 0xFF0066FF);
+    const QRgb disabledHighlight(isDarkSystemTheme ? 0xFF01060F : 0xFFF0F6FF);
+    const QRgb highlightedText(isDarkSystemTheme ? 0xFFFDFDFD : 0xFF090909);
+    const QRgb light(isDarkSystemTheme ? 0xFF1A1A1A : 0xFFF6F6F6);
+    const QRgb link(isDarkSystemTheme ? 0xFF2F86B1 : 0xFF45A7D7);
+    const QRgb mid(isDarkSystemTheme ? 0xFF626262 : 0xFFBDBDBD);
+    const QRgb midlight(isDarkSystemTheme ? 0xFF2C2C2C : 0xFFE4E4E4);
+    const QRgb text(isDarkSystemTheme ? 0xFFEFF0F2 : 0xFF353637);
+    const QRgb disabledText(isDarkSystemTheme ? 0x7FC8C9CB : 0x7F353637);
+    const QRgb shadow(isDarkSystemTheme ? 0xFF28282A : 0xFF28282A);
+    const QRgb toolTipBase(isDarkSystemTheme ? 0xFF000000 : 0xFFFFFFFF);
+    const QRgb toolTipText(isDarkSystemTheme ? 0xFFFFFFFF : 0xFF000000);
+    const QRgb window(isDarkSystemTheme ? 0xFF000000 : 0xFFFFFFFF);
+    const QRgb windowText(isDarkSystemTheme ? 0xFFD4D6D8 : 0xFF26282A);
+    const QRgb disabledWindowText(isDarkSystemTheme ? 0xFF3F4040 : 0xFFBDBEBF);
+    const QRgb placeholderText(isDarkSystemTheme ? 0x88C8C9CB : 0x88353637);
 
-    systemPalette.setColor(QPalette::ButtonText, QColor::fromRgba(0xFF26282A));
-    systemPalette.setColor(QPalette::Disabled, QPalette::ButtonText, QColor::fromRgba(0x4D26282A));
+    systemPalette.setColor(QPalette::Base, base);
+    systemPalette.setColor(QPalette::Disabled, QPalette::Base, disabledBase);
 
-    systemPalette.setColor(QPalette::BrightText, QColor::fromRgba(0xFFFFFFFF));
-    systemPalette.setColor(QPalette::Disabled, QPalette::BrightText, QColor::fromRgba(0x4DFFFFFF));
+    systemPalette.setColor(QPalette::Button, button);
 
-    systemPalette.setColor(QPalette::Dark, QColor::fromRgba(0xFF353637));
+    systemPalette.setColor(QPalette::ButtonText, buttonText);
+    systemPalette.setColor(QPalette::Disabled, QPalette::ButtonText, disabledButtonText);
 
-    systemPalette.setColor(QPalette::Highlight, QColor::fromRgba(0xFF0066FF));
-    systemPalette.setColor(QPalette::Disabled, QPalette::Highlight, QColor::fromRgba(0xFFF0F6FF));
+    systemPalette.setColor(QPalette::BrightText, brightText);
+    systemPalette.setColor(QPalette::Disabled, QPalette::BrightText, disabledBrightText);
 
-    systemPalette.setColor(QPalette::HighlightedText, QColor::fromRgba(0xFF090909));
+    systemPalette.setColor(QPalette::Dark, dark);
 
-    systemPalette.setColor(QPalette::Light, QColor::fromRgba(0xFFF6F6F6));
+    systemPalette.setColor(QPalette::Highlight, highlight);
+    systemPalette.setColor(QPalette::Disabled, QPalette::Highlight, disabledHighlight);
 
-    systemPalette.setColor(QPalette::Link, QColor::fromRgba(0xFF45A7D7));
+    systemPalette.setColor(QPalette::HighlightedText, highlightedText);
 
-    systemPalette.setColor(QPalette::Mid, QColor::fromRgba(0xFFBDBDBD));
+    systemPalette.setColor(QPalette::Light, light);
 
-    systemPalette.setColor(QPalette::Midlight, QColor::fromRgba(0xFFE4E4E4));
+    systemPalette.setColor(QPalette::Link, link);
 
-    systemPalette.setColor(QPalette::Text, QColor::fromRgba(0xFF353637));
-    systemPalette.setColor(QPalette::Disabled, QPalette::Text, QColor::fromRgba(0x7F353637));
+    systemPalette.setColor(QPalette::Mid, mid);
 
-    systemPalette.setColor(QPalette::Shadow, QColor::fromRgba(0xFF28282A));
+    systemPalette.setColor(QPalette::Midlight, midlight);
 
-    systemPalette.setColor(QPalette::ToolTipBase, QColor::fromRgba(0xFFFFFFFF));
-    systemPalette.setColor(QPalette::ToolTipText, QColor::fromRgba(0xFF000000));
+    systemPalette.setColor(QPalette::Text, text);
+    systemPalette.setColor(QPalette::Disabled, QPalette::Text, disabledText);
 
-    systemPalette.setColor(QPalette::Window, QColor::fromRgba(0xFFFFFFFF));
+    systemPalette.setColor(QPalette::Shadow, shadow);
 
-    systemPalette.setColor(QPalette::WindowText, QColor::fromRgba(0xFF26282A));
-    systemPalette.setColor(QPalette::Disabled, QPalette::WindowText, QColor::fromRgba(0xFFBDBEBF));
+    systemPalette.setColor(QPalette::ToolTipBase, toolTipBase);
+    systemPalette.setColor(QPalette::ToolTipText, toolTipText);
 
-    systemPalette.setColor(QPalette::PlaceholderText, QColor::fromRgba(0x88353637));
+    systemPalette.setColor(QPalette::Window, window);
+
+    systemPalette.setColor(QPalette::WindowText, windowText);
+    systemPalette.setColor(QPalette::Disabled, QPalette::WindowText, disabledWindowText);
+
+    systemPalette.setColor(QPalette::PlaceholderText, placeholderText);
 
     theme->setPalette(QQuickTheme::System, systemPalette);
 }
 
+void QQuickBasicTheme::updateTheme()
+{
+    QQuickTheme *theme = QQuickTheme::instance();
+    if (!theme)
+        return;
+    initialize(theme);
+}
+
 QT_END_NAMESPACE
diff --git x/qtdeclarative/src/quickcontrols/basic/qquickbasictheme_p.h y/qtdeclarative/src/quickcontrols/basic/qquickbasictheme_p.h
index bcedee2a12afdee2715ab01fd9d62d3f8e594beb..33a32afd9cc91433e14678e84099071bfe85bb0d 100644
--- x/qtdeclarative/src/quickcontrols/basic/qquickbasictheme_p.h
+++ y/qtdeclarative/src/quickcontrols/basic/qquickbasictheme_p.h
@@ -25,6 +25,7 @@ class Q_QUICKCONTROLS2BASIC_EXPORT QQuickBasicTheme
 {
 public:
     static void initialize(QQuickTheme *theme);
+    static void updateTheme();
 };
 
 QT_END_NAMESPACE
diff --git x/qtdeclarative/src/quickcontrols/basic/qtquickcontrols2basicstyleplugin.cpp y/qtdeclarative/src/quickcontrols/basic/qtquickcontrols2basicstyleplugin.cpp
index e2d726082b49b15ffc90b0012934dd45dfcfa00a..274891dc2ab3c952a50bc7c06bf26b12f8f627f5 100644
--- x/qtdeclarative/src/quickcontrols/basic/qtquickcontrols2basicstyleplugin.cpp
+++ y/qtdeclarative/src/quickcontrols/basic/qtquickcontrols2basicstyleplugin.cpp
@@ -21,6 +21,7 @@ public:
 
     QString name() const override;
     void initializeTheme(QQuickTheme *theme) override;
+    void updateTheme() override;
 
     QQuickBasicTheme theme;
 };
@@ -41,6 +42,11 @@ void QtQuickControls2BasicStylePlugin::initializeTheme(QQuickTheme *theme)
     this->theme.initialize(theme);
 }
 
+void QtQuickControls2BasicStylePlugin::updateTheme()
+{
+    this->theme.updateTheme();
+}
+
 QT_END_NAMESPACE
 
 #include "qtquickcontrols2basicstyleplugin.moc"
